### PR TITLE
contracts: Add automated weights for wasm instructions

### PR DESCRIPTION
**Pull request by athei**
This PR is the last piece of having a completely automatic weight determination for contracts that is able to generate secure weights. After adding weights for extrinsic and contract callable host functions this adds the weights for each supported instruction.

It also adds some additional static checks for contracts that are necessary to price all instructions securely:

- Number of globals
- Number of parameters a function can have
- Number of entries in the immediate value of BrTable

The resulting `Schedule` as benchmarked by the bot for the substrate runtime looks as follows:
<details>
<summary>Schedule (click to open)</summary>

```rust
Schedule {
    version: 0,
    enable_println: false,
    limits: Limits {
        event_topics: 4,
        stack_height: 512,
        globals: 256,
        parameters: 128,
        memory_pages: 16,
        table_size: 4096,
        br_table_size: 256,
        subject_len: 32,
        code_size: 524288,
    },
    instruction_weights: InstructionWeights {
        i64const: "1.6 ns",
        i64load: "157.3 ns",
        i64store: "226.3 ns",
        select: "6.1 ns",
        r#if: "7.9 ns",
        br: "3.0 ns",
        br_if: "6.3 ns",
        br_table: "14.6 ns",
        br_table_per_entry: "103 ps",
        call: "95.1 ns",
        call_indirect: "199.5 ns",
        call_indirect_per_param: "2.1 ns",
        local_get: "1.8 ns",
        local_set: "2.1 ns",
        local_tee: "2.1 ns",
        global_get: "6.6 ns",
        global_set: "10.7 ns",
        memory_current: "2.3 ns",
        memory_grow: "143.5 µs",
        i64clz: "2.3 ns",
        i64ctz: "2.3 ns",
        i64popcnt: "2.9 ns",
        i64eqz: "2.4 ns",
        i64extendsi32: "2.2 ns",
        i64extendui32: "2.2 ns",
        i32wrapi64: "2.3 ns",
        i64eq: "2.6 ns",
        i64ne: "2.5 ns",
        i64lts: "2.5 ns",
        i64ltu: "2.6 ns",
        i64gts: "2.5 ns",
        i64gtu: "2.5 ns",
        i64les: "2.5 ns",
        i64leu: "2.6 ns",
        i64ges: "2.6 ns",
        i64geu: "2.5 ns",
        i64add: "2.5 ns",
        i64sub: "2.6 ns",
        i64mul: "2.5 ns",
        i64divs: "8.2 ns",
        i64divu: "7.2 ns",
        i64rems: "8.1 ns",
        i64remu: "7.3 ns",
        i64and: "2.5 ns",
        i64or: "2.5 ns",
        i64xor: "2.6 ns",
        i64shl: "2.5 ns",
        i64shrs: "2.6 ns",
        i64shru: "2.6 ns",
        i64rotl: "2.5 ns",
        i64rotr: "2.6 ns",
    },
    host_fn_weights: HostFnWeights {
        caller: "3.7 µs",
        address: "3.7 µs",
        gas_left: "3.6 µs",
        balance: "8.1 µs",
        value_transferred: "3.6 µs",
        minimum_balance: "3.6 µs",
        tombstone_deposit: "3.6 µs",
        rent_allowance: "8.5 µs",
        block_number: "3.6 µs",
        now: "3.7 µs",
        weight_to_fee: "6.1 µs",
        gas: "1.9 µs",
        input: "7.8 µs",
        input_per_byte: "266 ps",
        r#return: "6.1 µs",
        return_per_byte: "676 ps",
        terminate: "678.6 µs",
        restore_to: "606.7 µs",
        restore_to_per_delta: "162.4 µs",
        random: "9.4 µs",
        deposit_event: "13.4 µs",
        deposit_event_per_topic: "132.4 µs",
        deposit_event_per_byte: "2.3 ns",
        set_rent_allowance: "10.0 µs",
        set_storage: "266.9 µs",
        set_storage_per_byte: "2.0 ns",
        clear_storage: "176.0 µs",
        get_storage: "35.7 µs",
        get_storage_per_byte: "1.5 ns",
        transfer: "185.9 µs",
        call: "130.2 µs",
        call_transfer_surcharge: "175.0 µs",
        call_per_input_byte: "505 ps",
        call_per_output_byte: "731 ps",
        instantiate: "495.1 µs",
        instantiate_per_input_byte: "1.5 ns",
        instantiate_per_output_byte: "737 ps",
        hash_sha2_256: "3.3 µs",
        hash_sha2_256_per_byte: "4.1 ns",
        hash_keccak_256: "3.3 µs",
        hash_keccak_256_per_byte: "3.3 ns",
        hash_blake2_256: "3.1 µs",
        hash_blake2_256_per_byte: "1.5 ns",
        hash_blake2_128: "3.1 µs",
        hash_blake2_128_per_byte: "1.5 ns",
    },
}
```
</details>

The strategy here is to include every wasm instruction on its own and **not** manually curate them into categories. Otherwise a manual intervention would be necessary whenever those categories don't fit the current wasm execution engine we are benchmarking those on. Our strategy minimizes assumptions and therefore maximizes the amount of automation. Creating a schedule for compiled execution (this schedule is for wasmi executed contracts) is possible with a simple re-benchmark without writing any new code.

However, there are some assumptions that we needed to make. They are explained here:
https://github.com/paritytech/substrate/blob/b3b6099535d75727c56b64e5d505f34d516e89fc/frame/contracts/src/schedule.rs#L112-L134

As always: The easiest way to review this might be commit by commit.

@jacogr This will also add changes to the `Schedule` data structure that therefore needs an update in its js mirror.

### CODE REVIEW

**Comment by athei** [comment]
cargo run --release --features runtime-benchmarks --manifest-path bin/node/cli/Cargo.toml -- benchmark --chain dev --steps 50 --repeat 100 --extrinsic "*" --execution=wasm --wasm-execution=compiled --output ./bin/node/runtime/src/weights --header ./HEADER --pallet pallet_contracts --heap-pages 4096

**Comment by athei** [comment]
/benchmark runtime custom --chain dev --steps 50 --repeat 100 --extrinsic "*" --execution=wasm --wasm-execution=compiled --output ./bin/node/runtime/src/weights --header ./HEADER --pallet pallet_contracts --heap-pages 4096

**Comment by athei** [comment]
/benchmark runtime custom --chain dev --steps 50 --repeat 100 --extrinsic "*" --execution=wasm --wasm-execution=compiled --output ./bin/node/runtime/src/weights --header ./HEADER --pallet pallet_contracts --heap-pages 4096

**Comment by athei** [comment]
/benchmark runtime custom --chain dev --steps 10 --repeat 20 --extrinsic "*" --execution=wasm --wasm-execution=compiled --output ./bin/node/runtime/src/weights --header ./HEADER --pallet pallet_contracts --heap-pages 4096

**Comment by athei** [comment]
I changed the benchmark for the `memory.grow` instruction. It result in too high costs for growing the memory because the benchmark did test with a contract that had a much larger memory than it is allowed for contracts according to the `Schedule`.

Costs for that instruction are down from `1.4 ms` to  `143.5 µs`. I updated the `Schedule` in the opening post with the current one (as added by the latest bot run).

This behaviour poses the question of **why** the run time of that instruction depends on how much memory is already allocated. @pepyakin suggested that it might be that *wasmi* uses a `Vec` instead of `mmap` to implement the linear memory because it could be build with the wrong features. However, I couldn't find evidence that this is the cause.

However, for the currently low amount of memory we allow we can live with those relatively high costs for memory growth. We might need to revisit this performance animality as soon as we want to increase the available memory.

File: frame/contracts/proc-macro/src/lib.rs
Comment (pepyakin) [inline frame/contracts/proc-macro/src/lib.rs:1]: The header?

File: frame/contracts/src/benchmarking/code.rs
Comment (athei) [reply frame/contracts/src/benchmarking/code.rs:177]: I did not really make this assumption. The benchmarks only require some uniform distribution each time. Not sure if I should, though.

Comment (athei) [reply frame/contracts/src/benchmarking/code.rs]: I changed it to use a specific rng engine. Should be reproducible now.

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [reply frame/contracts/src/benchmarking/mod.rs:2362]: I think xtask is overkill. But having some aliases for benchmarking related stuff would be really helpful in general.

File: frame/contracts/src/benchmarking/mod.rs
Comment (ascjones) [inline frame/contracts/src/benchmarking/mod.rs:1307]: :eyes:

**Review by ascjones — COMMENTED** [review]
Just a few typos after my first pass

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [reply frame/contracts/src/benchmarking/mod.rs:2362]: I put creating aliases in a `.cargo/config` to my TODO list. Closing here for now.

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [reply frame/contracts/src/benchmarking/mod.rs:2210]: Uhhh good catch. Copy paste bug.

File: frame/contracts/src/schedule.rs
Comment (athei) [reply frame/contracts/src/schedule.rs:129]: :D

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [reply frame/contracts/src/benchmarking/mod.rs:1307]: This is what I mean when I say we need to make assumptions about the address generation in #7482. Once that PR is merged I can remove this and also create a test from that benchmark using `create_test!`.

File: frame/contracts/src/schedule.rs
Comment (ascjones) [inline frame/contracts/src/schedule.rs:113]: `/// Describes the weight for all categories of supported wasm instructions.`

No longer accurate since weights per instruction now rather than categories

**Review by ascjones — APPROVED** [review]
All in all LGTM, pending a review from someone with more wasm expertise.

One observation: where the instruction costs are essentially equivalent (within a certain error threshold) could we equalize them e.g. all the `instr_i64*` are virtually identical `~24_650_000 + 7_400_000`, the minor differences being from the benchmarking run.

**Comment by athei** [comment]
> One observation: where the instruction costs are essentially equivalent (within a certain error threshold) could we equalize them e.g. all the `instr_i64*` are virtually identical `~24_650_000 + 7_400_000`, the minor differences being from the benchmarking run.

This is expected and most other contract platforms define manually curated categories of instructions in order to coalesce them. However, my goal here is to not make unnecessary assumptions about the execution engine. We just take the results at face value because changing the execution engine could break those category assumptions. We want that to be a highly automated process. So this category stuff won't fly here.

What you are proposing is a post process step to equalize weights that are closely together, right? I don't see the benefit of that. And frankly I don't want to introduce yet another tuning knob (what should the threshold be?).

File: frame/contracts/src/schedule.rs
Comment (athei) [reply frame/contracts/src/schedule.rs:113]: This is a new comment. Previously, we had one weight for all instructions. Not even categories. What we have are still categories (i64 and i32 versions of the same instructions are grouped together for example). The granularity is just much finer as what other platforms call "category".

File: frame/contracts/proc-macro/src/lib.rs
Comment (NikVolf) [inline frame/contracts/proc-macro/src/lib.rs:49]:
```suggestion
) -> proc_macro::TokenStream {
```

File: frame/contracts/proc-macro/src/lib.rs
Comment (NikVolf) [inline frame/contracts/proc-macro/src/lib.rs:87]:
```suggestion
#[cfg(not(feature = "full"))]
```

why not this?

File: frame/contracts/src/benchmarking/code.rs
Comment (NikVolf) [inline frame/contracts/src/benchmarking/code.rs:236]: are those ending in html docs somewhere?

if so, better to decorate with header separated with empty line

```suggestion
	/// Create a wasm module of `target_bytes` size. 
	
	/// Used to benchmark the performance of
	/// `put_code` for different sizes of wasm modules. The generated module maximizes
	/// instrumentation runtime by nesting blocks as deeply as possible given the byte budget.
```

File: frame/contracts/src/benchmarking/code.rs
Comment (NikVolf) [inline frame/contracts/src/benchmarking/code.rs:334]:
```suggestion
		}.into()
```

File: frame/contracts/src/benchmarking/code.rs
Comment (NikVolf) [inline frame/contracts/src/benchmarking/code.rs:347]:
```suggestion
		}.into()
```

File: frame/contracts/src/benchmarking/mod.rs
Comment (NikVolf) [inline frame/contracts/src/benchmarking/mod.rs:1635]: could just use `body::DynInstr::*` somewhere above

File: frame/contracts/src/benchmarking/mod.rs
Comment (NikVolf) [inline frame/contracts/src/benchmarking/mod.rs:1906]: this feels quite repetitive

can we make a function that creates `Sandbox` from some code?

File: frame/contracts/src/benchmarking/sandbox.rs
Comment (NikVolf) [inline frame/contracts/src/benchmarking/sandbox.rs:45]:
```suggestion
		let instance = Instance::new(&module.code, &env_builder, &mut ()).expect("Fail to create benchmarking Sandbox instance");
```

File: frame/contracts/src/wasm/mod.rs
Comment (NikVolf) [inline frame/contracts/src/wasm/mod.rs:1464]: wasn't the idea of exact numbers is that we don't commit something that can change gas cost and break consensus?

**Comment by NikVolf** [comment]
Sorry for being too late

Looks great, other few nits

File: frame/contracts/src/wasm/mod.rs
Comment (athei) [reply frame/contracts/src/wasm/mod.rs:1464]: I didn't know that. Given that only those two seemingly random tests had a check for an explicit gas costs I didn't thought there where some kind reason behind that. I would expected an explicit test exercising more than one functionality to check that the gas costs changes without changing the `Schedule`.

File: frame/contracts/src/benchmarking/code.rs
Comment (athei) [reply frame/contracts/src/benchmarking/code.rs:236]: The whole benchmarking module is private and guarded behind the `runtime-benchmarks` feature flag. It is extremely unlikely that anyone will consume those comments using html.

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [reply frame/contracts/src/benchmarking/mod.rs:1906]: Only the three benchmarks for the local instructions would profit from such a function. Every other benchmark needs some slight variation (setting some memory, globals, etc.). I don't think that this warrants a new function.

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [reply frame/contracts/src/benchmarking/mod.rs]: Makes sense. It is used a lot.

File: frame/contracts/src/benchmarking/mod.rs
Comment (NikVolf) [reply frame/contracts/src/benchmarking/mod.rs:1888]: You could use `sandbox(call_body).with_mem()` builder pattern in other cases, but it's up to you :)

**Review by NikVolf — APPROVED** [review]

**Comment by athei** [comment]
bot merge

**Pull request merged**
